### PR TITLE
Fix scope propagation for multiple specs #9175

### LINKED ISSUE #9175: Scope is not propagated from layer to specs in test suites.
For various historical reasons (migration from a ZIO1 project), I need all my tests to run with `ClockLive` and `RandomLive`.  This is surprisingly hard to do in ZIO2.  Ideally, a layer like this:

```scala
  def realClock: ULayer[Unit] = ZLayer.scoped {
    for {
      _ <- ZIO.withClockScoped(Clock.ClockLive)
      _ <- ZIO.withRandomScoped(RandomLive)
    } yield ()
  }
```
Should be enough to override the default choice of the test environment.  Let’s put that in practice:

```scala
object RealClockSpec extends ZIOSpec[Unit] {
  val spec = suite("full")(single, singleWithoutProvide, aSuite) @@ sequential

  // this DOES work with a single test!
  def single =
    oneTest("single").provideLayerShared(realClock)

  // This does not work with a suite
  def aSuite = suite("foo")(
    oneTest("1"),
    oneTest("2")
  ).provideSomeLayerShared(realClock) @@ sequential

  // this does not work (scope not inherited from the bootstrap)
  def singleWithoutProvide = oneTest("singleWithoutProvide")

  // the scope of the layer does not propagate to the specs
  override val bootstrap: ULayer[Unit] = realClock

  def oneTest(s: String): Spec[Any, Nothing] = test(s) {
    ZIO.scoped {
      for {
        // TestRandom.deterministic produces  -1295463240
        x   <- Random.nextInt
        now <- Clock.localDateTime 
        // now will be 1970 under the TestClock
        _   <- ZIO.logInfo(s"id is $x $now")
      } yield assertTrue(x != -1295463240)
    }
  }

  def realClock: ULayer[Unit] = ZLayer.scoped {
    for {
      _ <- ZIO.withClockScoped(Clock.ClockLive)
      _ <- ZIO.withRandomScoped(RandomLive)
    } yield ()
  }
}
```

It’s a bit hard to swallow that `provideLayerShared` has a different semantic when it applies to a suite or to a single test.  So we have at least one  bug, don’t we ?  

The environment computed by layers is correctly propagated, of course.  Otherwise the bug would be much more visible.   But the `Scope` defined by the `realClock` layer does not appear  to propagate properly, except in the trivial `SpecCase::TestCase`

The code [here](https://github.com/zio/zio/blob/e41d9c5fd46d724ec1b435d10d68903a74932acb/test/shared/src/main/scala/zio/test/Spec.scala#L343) looks ok.  We observe the environment computed once, and provided to the list of specs.  There is an attempt to define a common Scope for all the specs. So maybe the issue is in the runner ?

## Related issue
There is a slightly related behaviour that presumably has the same root cause.  I don’t know if it deserves a different ticket.  Maybe it’s even consider a correct behavior, after all.

We certainly have some anti-pattern in our test suites.  Among the things we do, is having sequential dependant specs like
```scala
   val spec = suite(
      writeSomethingToDB(),
      writeSomethingElseToDB(),
      testSomethingInDB()
    ). provideSomeLayerShared(dbLayer) @@ sequential
```
I don’t know if it’s good or bad practice, but we have tons of tests like that.  The `TestRandom.deterministic` state is reset before each spec and that does not feel right, does it? The consequence is that  we generate the same ID for the DB writes and cause spurious insertion failures.  
This is not a surprising behaviour given that the state of the `TestRandom.deterministic` is kept in the current scope and the scope is lost, as previously observed.

In fact, I could live with `TestRandom.deterministic` if it would not reset between individual specs.  But I can understand that this is the expected behavior.

### PR DESCRIPTION
**Pull request by clarzte**
/claim #9175

I found the commit which introduced the bug https://github.com/zio/zio/commit/ff01fa4994597391d32ed373bf9f18fd93d61d76#diff-ad3f1b449cac121bd82f6d24a2a105e90334bc80deaf7a1fd050ab950b821ac5R345

I don't fully understand why the current implementation fails, so I just reverted the code to when it worked.

### CODE REVIEW

File: test-tests/shared/src/test/scala/zio/test/SpecSpec.scala [hunk 1]
```diff
@@ -78,6 +78,24 @@ object SpecSpec extends ZIOBaseSpec {
         for {
           summary <- execute(spec)
         } yield assertTrue(summary.success == 1)
+      },
+      test("propagates the scope to multiple tests") {
+        def assertion = ZIO.scoped {
+          for {
+            value <- Random.nextInt
+          } yield assertTrue(value != -1295463240) // TestRandom.deterministic produces -1295463240
```
Comment (kyri-petrou) [inline test-tests/shared/src/test/scala/zio/test/SpecSpec.scala]: I think it's better to have a more concrete check for this, e.g.,:

```scala
          for {
            rnd <- ZIO.random
          } yield assertTrue(rnd == Random.RandomLive)
```

File: test-tests/shared/src/test/scala/zio/test/SpecSpec.scala [hunk 2]
```diff
@@ -78,6 +78,24 @@ object SpecSpec extends ZIOBaseSpec {
...
+        }
+        val spec = suite("suite")(
+          test("test1") {
+            assertion
+          },
+          test("test2") {
+            assertion
+          }
+        ).provideLayerShared(ZLayer.scoped(ZIO.withRandomScoped(Random.RandomLive)))
```
Comment (kyri-petrou) [inline test-tests/shared/src/test/scala/zio/test/SpecSpec.scala:123]: We also need to test that `provideSomeLayerShared` works as expected. If it doesn't we need to fix it as well

Comment (clarzte) [reply test-tests/shared/src/test/scala/zio/test/SpecSpec.scala:123]: good catch, done

File: test-tests/shared/src/test/scala/zio/test/SpecSpec.scala [hunk 3]
```diff
@@ -199,6 +245,24 @@ object SpecSpec extends ZIOBaseSpec {
           }
         ).provideSomeLayerShared[Scope](layer).provideLayer(Scope.default) @@ sequential
         assertZIO(succeeded(spec))(isTrue)
+      },
+      test("propagates the scope to multiple tests") {
```
Comment (kyri-petrou) [inline test-tests/shared/src/test/scala/zio/test/SpecSpec.scala:249]: This title is reused, which in the past I found that it doesn't play very nice in some cases. Also in zio-test, it's generally more idiomatic to structure tests hierarchically, e.g.,

```scala
suite("propagates the scope to multiple tests")(
  test("provideLayerShared") { ... },
  test("provideSomeLayerShared") { ... }
)
```

**Review by kyri-petrou — COMMENTED** [review]
LGTM, just a minor comment

File: test-tests/shared/src/test/scala/zio/test/SpecSpec.scala (hunk 3 above)
Comment (kyri-petrou) [reply test-tests/shared/src/test/scala/zio/test/SpecSpec.scala:249]: Oops, my bad, I just realised they're already scoped. Please disregard my comment

**Review by kyri-petrou — APPROVED** [review]

**Pull request merged**